aten/src/ATen/native/SpectralOps.cpp

```diff
@@ -323,49 +436,97 @@ Tensor fft_irfftn(const Tensor& self, c10::optional<IntArrayRef> s,
+    return at::_fft_c2r_out(out, x, desc.dim, static_cast<int64_t>(norm), last_dim_size);
+  } else {
+    return at::_fft_c2r(x, desc.dim, static_cast<int64_t>(norm), last_dim_size);
+  }
+}
+
+Tensor fft_irfftn(const Tensor& self,
+                  c10::optional<IntArrayRef> s,
+                  c10::optional<IntArrayRef> dim,
+                  c10::optional<std::string> norm_str) {
+  return fft_irfftn_impl({}, self, s, dim, norm_str);
+}
+
+Tensor& fft_irfftn_out(Tensor& out, const Tensor& self,
+                       c10::optional<IntArrayRef> s,
+                       c10::optional<IntArrayRef> dim,
+                       c10::optional<std::string> norm_str) {
+  fft_irfftn_impl(out, self, s, dim, norm_str);
+  return out;
 }
 
 Tensor fft_fft2(const Tensor& self, c10::optional<IntArrayRef> s,
                 IntArrayRef dim, c10::optional<std::string> norm) {
   return native::fft_fftn(self, s, dim, std::move(norm));
 }
 
+Tensor& fft_fft2_out(Tensor& out, const Tensor& self, c10::optional<IntArrayRef> s,
+                     IntArrayRef dim, c10::optional<std::string> norm) {
+  return native::fft_fftn_out(out, self, s, dim, std::move(norm));
+}
+
 Tensor fft_ifft2(const Tensor& self, c10::optional<IntArrayRef> s,
                 IntArrayRef dim, c10::optional<std::string> norm) {
   return native::fft_ifftn(self, s, dim, std::move(norm));
 }
 
+Tensor& fft_ifft2_out(Tensor& out, const Tensor& self, c10::optional<IntArrayRef> s,
+                      IntArrayRef dim, c10::optional<std::string> norm) {
+  return native::fft_ifftn_out(out, self, s, dim, std::move(norm));
+}
+
 Tensor fft_rfft2(const Tensor& self, c10::optional<IntArrayRef> s,
                 IntArrayRef dim, c10::optional<std::string> norm) {
   return native::fft_rfftn(self, s, dim, std::move(norm));
 }
 
+Tensor& fft_rfft2_out(Tensor& out, const Tensor& self, c10::optional<IntArrayRef> s,
+                      IntArrayRef dim, c10::optional<std::string> norm) {
+  return native::fft_rfftn_out(out, self, s, dim, std::move(norm));
+}
+
 Tensor fft_irfft2(const Tensor& self, c10::optional<IntArrayRef> s,
                   IntArrayRef dim, c10::optional<std::string> norm) {
   return native::fft_irfftn(self, s, dim, std::move(norm));
 }
 
-Tensor fft_fftfreq(int64_t n, double d, const TensorOptions& options) {
-  ScalarType dtype = typeMetaToScalarType(options.dtype());
+Tensor& fft_irfft2_out(Tensor& out, const Tensor& self, c10::optional<IntArrayRef> s,
+                       IntArrayRef dim, c10::optional<std::string> norm) {
+  return native::fft_irfftn_out(out, self, s, dim, std::move(norm));
+}
+
+Tensor& fft_fftfreq_out(Tensor& out, int64_t n, double d) {
+  ScalarType dtype = out.scalar_type();
   TORCH_CHECK(at::isFloatingType(dtype) || at::isComplexType(dtype),
               "fftfreq requires a floating point or complex dtype");
```

Comment:
What are the error messages for other tensor creation ops, like `torch.arange` and `torch.zeros`, like for the out vs tensor options case? What happens if both out= and some tensor options are specified?

Reply:
>What happens if both out= and some tensor options are specified?

The function I moved out of `python_torch_functions.cpp` is automatically called by codegen to ensure the tensor options arguments are valid. For some reason this happens even though I don't take any `TensorOptions` arguments here.

Reply:
That's odd. Don't other functions, like those mentioned above, have this same issue? What's the implication of this function always being called even though no TensorOptions are in the signature?

Reply:
Or are the `fftfreq` functions the first tensor factories to hit some unique aspect of this issue?

Reply:
>What's the implication of this function always being called even though no TensorOptions are in the signature?

The non-out variant uses `TensorOptions` to create the out tensor; the out variant respects the `TensorOptions` of the `out` tensor.  And the python API ensures that if both are given, they must be consistent.

>Or are the fftfreq functions the first tensor factories to hit some unique aspect of this issue?

I don't think it's unique, I think it's just a quirk of the python codegen. For example, I see that `empty.out` doesn't have a `dtype` argument yet I can still do `torch.empty(100, out=out, dtype=torch.float)`.

There is also no `TensorOptions` argument in the native function:
https://github.com/pytorch/pytorch/blob/4883d39c6fd38431bdc60e1db6402e251429b1e1/aten/src/ATen/native/TensorFactories.cpp#L198-L201

Reply:
That's really interesting. cc @bhosmer for yet another quirk of our current codegen.